### PR TITLE
Upstream `argmax` shape function.

### PR DESCRIPTION
**Pull request by silvasean**
Keeping this first commit simple to test out the flow. Will bulk-add the
rest once this one goes through.

Shape function taken from:
https://github.com/llvm/torch-mlir/blob/5192a4e9f3e4cbc77838b70153cd4632aa43dd7f/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py#L488

### CODE REVIEW

File: torch/jit/_shape_functions.py [hunk 1]
```diff
@@ -837,6 +837,22 @@ def flatten(input: List[int], start_dim: int, end_dim: int):
         shape.append(input[i])
     return shape
 
+def _reduce_along_dim(self: List[int], dim: int, keepdim: bool):
```
Comment (eellison) [inline torch/jit/_shape_functions.py:840]: So, the way that this works is that we compile these in torchscript then write the serialized functions out to a cpp file so the shape functions will work without a python dependency.. this process depends on torchscript being already built, so unfortunately we can automate this in the build step. 

You need to run `cd ~/pytorch/ && python torchgen/shape_functions/gen_jit_shape_functions.py` after making the changes here. 

Also, there was a recent moving of these files so it needs a small change in order to work:

```
diff --git a/torchgen/shape_functions/gen_jit_shape_functions.py b/torchgen/shape_functions/gen_jit_shape_functions.py
index cd43843600..d25539d3fa 100644
--- a/torchgen/shape_functions/gen_jit_shape_functions.py
+++ b/torchgen/shape_functions/gen_jit_shape_functions.py
@@ -114,7 +114,7 @@ def write_decomposition_util_file(path: str) -> None:
 
 
 def main() -> None:
-    pytorch_dir = Path(__file__).resolve().parents[3]
+    pytorch_dir = Path(__file__).resolve().parents[2]
     upgrader_path = pytorch_dir / "torch" / "csrc" / "jit" / "runtime"
     write_decomposition_util_file(str(upgrader_path))
```

File: torch/jit/_shape_functions.py [hunk 2]
```diff
@@ -837,6 +837,22 @@ def flatten(input: List[int], start_dim: int, end_dim: int):
...
+    dim = maybe_wrap_dim(dim, len(self))
+    out: List[int] = []
+    for i, self_dim in enumerate(self):
+        if i == dim:
+            if keepdim:
```
Comment (eellison) [inline torch/jit/_shape_functions.py:845]: Can you add `assert_jit_shape_analysis` to `
```
ReductionOpInfo(
        'argmax',
        supports_multiple_dims,
        ...
)
```
in 
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py

wont let me link bc the file is too large 😢

**Review by eellison — COMMENTED** [review]
Cool! one comment about using `aten::`, and then the upgrader needs a small fix. 

I will submit a PR to make specifying the schema a bit more ergonomic, and I can also add a runtime hook to register these so you can test out the new ones added without rebuilding..

**Comment by eellison** [comment]
cc @Gamrix

File: torch/jit/_shape_functions.py (hunk 1 above)
Comment (Gamrix) [reply torch/jit/_shape_functions.py:840]: That script will not work until this diff is landed and you rebase onto it. 
https://github.com/pytorch/pytorch/pull/76703

**Review by eellison — APPROVED** [review]
sweet! lgtm! thank you!

**Pull request closed without merging**